### PR TITLE
feat: Add a featured sort option for CatalogItems and paginate

### LINKED ISSUE #4983: Tags - Backend: Tag Listing Page products can be sorted by Featured order
## Story

After an operator user features products for a tag, the tag listing page should be able to display the products in that featured order.

### LINKED ISSUE #5004: OP2.0 Tags table pagination issues
## Issue Description
The previous button, when enabled, causes the pagination to break.

The next button becomes unresponsive, and the previous button allows you to go beyond the top end of the list and get a page with less than 10 results when you should have at least 10 on the first page with a large enough dataset.

## Steps to Reproduce

1. Using our sample data or, many many tags added to http://localhost:3000/operator/tags
2. Click the `next` button
3. Click the now enabled `previous` button
    1. See the previous button is still enabled even though you should be on the `first` page
4. Click the `next` button again
    1. The next button does nothing when it should take you to the next page
5. Click the `previous` button again
    1. You are now taken the the frist page again? but now with less than `10` results
    2. The previous button will be disabled now
    3. The next button does not take you to the next page of results

## Possible Solution

- Investigate the pagination helper util used in the TagDataTable component
- Investigate the GraphQL resolver to see if there is some issue there
- Investigate the data source so see if maybe that is causing the issue somehow

### Versions
Node: 10.7.0
NPM: 6.5.0
Meteor Node: 8.11.4
Meteor NPM: 6.4.1
Reaction CLI: 0.29.0
Reaction: 2.0.0-rc.9
Reaction branch: develop
Docker: 18.09.1

### PR DESCRIPTION
**Pull request by machikoyasuda**
closes #4983 
closes #5004

Currently, the `catalogItems` query returns a tag's products in order by newest products first. Now that the operator tag management tool is finished and deployed, a developer should be able to query for a tag's `catalogItems` in the order set in a `tag`. This ticket modifies the query to add a `featured` as a sorting option, and also makes `featured` the default option.

## Technical specs
- adds a `featured` to `CatalogItemSortByField`
- adds a `catalogItemsAggregate()` method to the `catalogItems()` resolver that returns `catalogItems` in the order of its tag's `featuredProductIds`, passed through to `getPaginatedAggregateResponse`
- fixes `getPaginatedAggregateResponse()` and `applyPaginationToMongoAggregation()` to so that `catalogItems` and `productsByTagId` queries paginate correctly
- adds integration specs ✨ for `catalogItems` query
- adds integration specs ✨ for `productsByTagId` query

## Test cases:
- Test tag with fewer than 20 products
- Test tag with more than 20 products, on first page
- Test tag with more than 20 products, on middle pages
- Test tag with more than 20 products, on last page
- Test tag with more than 20 products, going back to first page using `after`
- Test tag with 0 featuredProductIds
- Test on StarterKit
- Test on Admin

## Testing

After this ticket is merged, https://github.com/reactioncommerce/reaction-next-starterkit/pull/498 can be tested and merged to display the products in the Starter-Kit with this query with the `featured` sorting option. This Starter Kit PR can also be used to test both the front-end and back-end:

1. Log in as Admin
2. Go to http://localhost:3000/operator/tags/
3. Choose a tag to edit http://localhost:3000/operator/tags/edit/cmVhY3Rpb24vdGFnOnJwakN2VEJHamhCaTJ4ZHJv
4. Make the last product show up first by adding `0` to the Priority column. Save.
5. View the tag on http://localhost:4000/tag/shop

Test on GraphQL with a query variable: `"slugOrId": "shop"`.

First, confirm a Tag's `featuredProductIds` with this query for a tag, 

```gql
query getTagId($slugOrId: String!) {
  tag(slugOrId:$slugOrId) {
    _id
    name
    featuredProductIds
    displayTitle
    slug
  }
}
```

You should get something like:

```gql
{
  "data": {
    "tag": {
      "_id": "cmVhY3Rpb24vdGFnOnJwakN2VEJHamhCaTJ4ZHJv",
      "name": "Shop",
      "featuredProductIds": [
        "J4FquyjPnvTyw6kQA",
        "BCTMZ6HTxFSppJESk",
        "i7mbZ62dcwtnN8chF"
      ],
      "displayTitle": "Test Display Title",
      "slug": "shop"
    }
  }
}
```

Then, pass `featured` into the `catalogItems` query:

```gql
query catalogItems($shopIds: [ID]!, $tagIds: [ID]) {
  catalogItems(shopIds: $shopIds, tagIds:$tagIds, sortBy:featured) {
    totalCount
    pageInfo {
      hasNextPage
      hasPreviousPage
    }
    nodes {
      ... on CatalogItemProduct {
        product {
          title
          slug
        }
      }
    }
  }
}
```

This query should return the products in order of `featuredProductIds`.

## Example

1. Featured `Croakies` and `One by Optic` to the 0 and 1 spots in Admin:
screen shot 2019-02-14 at 12 33 57 pm

2. Sorted by `featured-desc` in Starter Kit:
screen shot 2019-02-14 at 12 33 43 pm

### CODE REVIEW

**Comment by machikoyasuda** [comment]
@mikemurray I'll update the Admin tool ordering in a second PR.

File: imports/plugins/core/catalog/server/no-meteor/resolvers/Query/catalogItems.js [hunk 1]
```diff
@@ -10,23 +10,35 @@ import { decodeTagOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/tag
...
+  if (connectionArgs.sortBy === "featured") {
+    if (tagIds.length === 0) {
+      throw new Error("A tag ID is required.");
+    }
+    const tagId = tagIds[0];
+    const query = await context.queries.catalogItemsAggregate(context, {
+      shopIds,
+      tagId
+    });
+    return getPaginatedAggregateResponse(query, connectionArgs);
```
[the remark this answers is not in the document]
Comment (mikemurray) [reply imports/plugins/core/catalog/server/no-meteor/resolvers/Query/catalogItems.js]: correction: `last` does **NOT** seem to start from the end of the list

**Comment by machikoyasuda** [comment]
@mikemurray Admin query looks good 👍 
 

screen shot 2019-02-14 at 3 36 36 pm

File: imports/plugins/core/tags/server/no-meteor/queries/productsByTagId.js [hunk 2]
```diff
@@ -50,16 +50,32 @@ export default async function productsByTagId(context, params) {
...
+  // Projection: Add a featuredPosition by order
+  const projection = {
+    $project: {
+      _id: 1,
+      position: 1,
+      title: 1,
+      sortPosition: {
+        $cond: {
+          if: { $lt: ["$position", 0] },
+          then: { $add: [{ $abs: "$position" }, positions.length] },
```
Comment (machikoyasuda) [inline imports/plugins/core/tags/server/no-meteor/queries/productsByTagId.js]: `positions` will be null if a Tag doesn't have any `featuredProductIds` yet, and `positions.length` will fail with: ` "Cannot read property 'length' of undefined",`

File: imports/plugins/core/graphql/server/no-meteor/util/applyPaginationToMongoAggregation.js [hunk 3]
```diff
@@ -1,90 +1,84 @@
...
 
-  const results = await collection.aggregate([...pipeline, facet]).toArray();
+  const unpaginatedResults = await collection.aggregate([...pipeline, facet]).toArray();
+  let hasPreviousPage;
+  let hasNextPage;
+  let paginatedItems;
+  let totalCount;
+  const limit = first || last || DEFAULT_LIMIT;
+
+  if (unpaginatedResults[0].nodes.length === 0) {
```
Comment (rosshadden) [inline imports/plugins/core/graphql/server/no-meteor/util/applyPaginationToMongoAggregation.js]: Do you know at these points that `unpaginatiedResults` has at least one item?

**Comment by machikoyasuda** [comment]
Thank youuuu @rosshadden + @mikemurray + @aldeed !

**Pull request merged**